pkg/steps/run_test.go

```diff
@@ -40,135 +46,289 @@ func (*fakeStep) Objects() []ctrlruntimeclient.Object { return nil }
-		}
-	}
-}
+	for _, tc := range testCases {
+		t.Run(tc.id, func(t *testing.T) {
+			shouldFail := 0
+			shouldRun := 0
 
-func TestRunFailureCase(t *testing.T) {
-	root := &fakeStep{
-		name:      "root",
-		shouldRun: true,
-		requires:  []api.StepLink{api.ExternalImageLink(api.ImageStreamTagReference{Namespace: "ns", Name: "base", Tag: "latest"})},
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceRoot)},
-	}
-	other := &fakeStep{
-		name:      "other",
-		shouldRun: true,
-		requires:  []api.StepLink{api.ExternalImageLink(api.ImageStreamTagReference{Namespace: "ns", Name: "base", Tag: "other"})},
-		creates:   []api.StepLink{api.InternalImageLink("other")},
-	}
-	src := &fakeStep{
-		name:      "src",
-		shouldRun: true,
-		requires:  []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceRoot)},
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceSource)},
-	}
-	bin := &fakeStep{
-		name:      "bin",
-		shouldRun: true,
-		requires:  []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceSource)},
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceBinaries)},
-	}
-	testBin := &fakeStep{
-		name:      "testBin",
-		shouldRun: true,
-		requires:  []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceSource)},
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceTestBinaries)},
-	}
-	rpm := &fakeStep{
-		name:      "rpm",
-		runErr:    errors.New("oopsie"),
-		shouldRun: true,
-		requires:  []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceBinaries)},
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceRPMs)},
-	}
-	unrelated := &fakeStep{
-		name:      "unrelated",
-		shouldRun: false,
-		requires:  []api.StepLink{api.InternalImageLink("other"), api.InternalImageLink(api.PipelineImageStreamTagReferenceRPMs)},
-		creates:   []api.StepLink{api.InternalImageLink("unrelated")},
-	}
-	final := &fakeStep{
-		name:      "final",
-		shouldRun: false,
-		requires:  []api.StepLink{api.InternalImageLink("unrelated")},
-		creates:   []api.StepLink{api.InternalImageLink("final")},
-	}
+			var steps []api.Step
+			for _, step := range tc.steps {
+				if step.runErr != nil {
+					shouldFail++
+				}
+				if step.shouldRun {
+					shouldRun++
+				}
+				steps = append(steps, step)
+			}
 
-	suites, _, err := Run(context.Background(), api.BuildGraph([]api.Step{root, other, src, bin, testBin, rpm, unrelated, final}))
-	if err == nil {
-		t.Error("got no error but expected one")
-	}
-	if suites.Suites != nil && len(suites.Suites) != 1 || len(suites.Suites[0].TestCases) != 6 || suites.Suites[0].NumTests != 6 || suites.Suites[0].NumFailed != 1 {
-		t.Errorf("unexpected junit output: %#v", suites.Suites[0])
-	}
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
```

Comment:
Won't this double-cancel for cases that don't need a cancel? Is that not an error?

Reply:
This will run after we fail or succeed. In other words, we don't need it, but linter complains :)